src/QsCompiler/TextProcessor/ParsingPrimitives.fs

```diff
@@ -32,12 +32,25 @@ let private buildQsExpression kind (range : Position * Position) = (kind, range)
 let private buildQsSymbol     kind (range : Position * Position) = (kind, range) |> QsSymbol.New
 let private buildQsType       kind (range : Position * Position) = (kind, range) |> QsType.New
 
-/// returns true if the given char is a valid symbol start - i.e. if it is an ascii letter or an underscore
-let internal isSymbolStart c = isAsciiLetter c || c = '_'
-
-/// returns true if the given char is a valid symbol continuation - i.e. if it is an ascii letter, a digit, or an underscore
-let internal isSymbolContinuation c = isAsciiLetter c || isDigit c || c = '_'
+/// returns true if the given char is a valid start symbol for an identifier: Underscore or a Unicode character of classes Lu, Ll, Lt, Lm, Lo, Nl.
+/// https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#identifiers
+let internal isSymbolStart c =
+    c = '_' ||
+    System.Char.IsLetter(c) || // Covers Lu, Ll, Lt, Lm, Lo
+    System.Char.GetUnicodeCategory(c) = System.Globalization.UnicodeCategory.LetterNumber // Nl
 
+/// returns true if the given char is a valid part symbol for an identifier: Unicode character of classes Lu, Ll, Lt, Lm, Lo, Nl; Nd; Pc; Mn, Mc; Cf.
+/// https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#identifiers
+let internal isSymbolContinuation c =
+    System.Char.IsLetter(c) || // Covers Lu, Ll, Lt, Lm, Lo
```

Comment:
Do you have any idea how expensive System.Char.GetUnicodeCategory is? 
Since this is used a lot during processing, depending on that it may possibly make sense to split out "c ='_'" as a separate case. I suppose all the other checks like e.g. whether c is a whitespace character are not cheaper to split out. But I haven't looked in detail what is quick and what isn't.

Reply:
I will do some tests in C# - this is calling same framework functions. I wouldn't expect GetUnicodeCategory to be too expensive - it is likely checking ranges.

---
In reply to: [442521582](https://github.com/microsoft/qsharp-compiler/pull/467#discussion_r442521582) [](ancestors = 442521582)

Reply:
A revelation: char.IsLetter() calls into the same function to get Unicode category as char.GetUnicodeCategory does. So char.IsLetter is slower. GetUnicodeCategory consists of 3 table clever lookups. It's only 2.5 times slower than this FAKE code:
                if (char.IsLetter(c)) {
                    result = System.Globalization.UnicodeCategory.NonSpacingMark;
                } else {
                    result = System.Globalization.UnicodeCategory.SpacingCombiningMark;
                }
I believe that there's no need to split c='_'.

---
In reply to: [442521582](https://github.com/microsoft/qsharp-compiler/pull/467#discussion_r442521582) [](ancestors = 442521582)

Reply:
Thanks for looking into this!